zookeeper/src/main/java/com/linecorp/armeria/client/endpoint/zookeeper/ZookeeperEndpointGroup.java

```diff
@@ -0,0 +1,175 @@
+                endpointList = nodeValueToEndpointList.convert(nodeData);
+            }
+
+        } catch (Exception e) {
+            throw new EndpointGroupException("failed to connect to ZooKeeper:" + zkConnectionStr);
+        }
+    }
+
+    /**
+     * Create a Zookeeper endpoint group with a {@link DefaultNodeValueConverter}.
+     * @param zkConnectionStr  A connection string containing a comma separated list
+     *        of host:port pairs,each corresponding to a ZooKeeper server
+     * @param zNode a zNode path e.g. {@code "/groups/productionGroups"}
+     * @param sessionTimeout   Zookeeper session timeout in milliseconds
+     */
+    public ZookeeperEndpointGroup(String zkConnectionStr, String zNode, int sessionTimeout) {
+        this(zkConnectionStr, zNode, sessionTimeout, new DefaultNodeValueConverter());
+    }
+
+    @Override
+    public List<Endpoint> endpoints() {
+        return this.endpointList;
+
+    }
+
+    public class ZkWatcher implements Watcher, StatCallback {
+        @Override
+        public void process(WatchedEvent event) {
+            String path = event.getPath();
+            if (event.getType() != Event.EventType.None) {
+                if (path != null && path.equals(znode)) {
+                    // Something has changed on the node, let's find out
+                    zooKeeper.exists(znode, true, this, null);
+                }
+            }
+
+        }
+
+        @Override
+        public void processResult(int responseCodeInt, String path, Object ctx, Stat stat) {
+            boolean exists;
+            Code responseCode = get(responseCodeInt);
+            switch (responseCode) {
+                case OK:
+                    exists = true;
+                    break;
+                case NONODE:
+                    exists = false;
+                    break;
+                case SESSIONEXPIRED:
+                case NOAUTH:
+                    return;
+                default:
+                    // Retry errors
+                    zooKeeper.exists(znode, true, this, null);
+                    return;
+            }
+
+            byte[] nodeData = null;
+            if (exists) {
+                try {
+                    nodeData = zooKeeper.getData(znode, false, null);
+                } catch (KeeperException e) {
+                    e.printStackTrace();
+                } catch (InterruptedException e) {
+                    return;
+                }
+            }
+            if ((nodeData == null && nodeData != prevData) ||
+                (nodeData != null && !Arrays.equals(prevData, nodeData))) {
+                prevData = nodeData;
+                ZookeeperEndpointGroup.this.endpointList = nodeValueToEndpointList.convert(prevData);
+            }
+        }
+    }
+
+    /**
+     *  Close the under Zookeeper connection.
+     * @throws Exception Closing exception
+     */
+    public void close() throws Exception {
+        zooKeeper.close();
```

Comment:
Given that we don't really have anything to do to recover from this failure, could we just log and now throw any exceptions?